GiniVision/Classes/ToolTipView.swift

```diff
@@ -0,0 +1,303 @@
+                assertionFailure("The tip cannot be shown outside the super view")
+            } else {
+                x = referenceViewAbsoluteFrame.origin.x  - size.width
+                y = referenceViewAbsoluteFrame.origin.y - margin.top
+            }
+        }
+        
+        if x < 0 || superview.frame.width - x < size.width {
+            x = superview.frame.width - size.width
+        }
+        
+        if superview.frame.height - y < size.height {
+            y = referenceViewAbsoluteFrame.origin.y + referenceViewAbsoluteFrame.height - size.height
+        }
+        
+        self.frame = CGRect(origin: CGPoint(x: x, y: y), size: size)
+    }
+    
+    fileprivate func arrangeArrow(withSuperView superView:UIView?) {
+        guard let referenceViewAbsoluteFrame = absoluteFrame(for: referenceView, inside: superView) else { return }
+        
+        let x:CGFloat
+        let y:CGFloat
+        switch toolTipPosition {
+        case .above:
+            x = referenceViewAbsoluteFrame.origin.x + referenceView.frame.width / 2 - self.frame.origin.x - arrowView.frame.width / 2
+            y = tipContainer.frame.height + tipContainer.frame.origin.y
+        case .below:
+            x = referenceViewAbsoluteFrame.origin.x + referenceView.frame.width / 2 - self.frame.origin.x - arrowView.frame.width / 2
+            y = 0
+        case .left:
+            x = tipContainer.frame.width + tipContainer.frame.origin.x
+            y = referenceViewAbsoluteFrame.origin.y + referenceView.frame.height / 2 - self.frame.origin.y - arrowView.frame.width / 2
+        case .right:
+            x = 0
+            y = referenceViewAbsoluteFrame.origin.y + referenceView.frame.height / 2 - self.frame.origin.y - arrowView.frame.width / 2
+        }
+        arrowView.frame.origin = CGPoint(x: x, y: y)
+    }
+    
+    // MARK: Actions
+    @objc fileprivate func closeAction() {
+        self.dismiss(withCompletion: nil)
+    }
+    
+    // MARK: Frame and size calculations
+    fileprivate func size(forText text: String, withFont font:UIFont) -> CGSize {
+        let attributes = [NSFontAttributeName : font]
+        
+        var textSize = text.boundingRect(with: CGSize(width: textWidth, height: CGFloat.greatestFiniteMagnitude), options: NSStringDrawingOptions.usesLineFragmentOrigin, attributes: attributes, context: nil).size
+        
+        textSize.width = textWidth
+        textSize.height = ceil(textSize.height)
+        
+        return textSize
+    }
+    
+    fileprivate func absoluteFrame(for view:UIView, inside superView:UIView?) -> CGRect? {
+        guard let superView = superView, let referenceViewParent = referenceView.superview else { return nil }
+        
+        return referenceViewParent.convert(referenceView.frame, to: superView)
+    }
+    
+    // MARK: Constraints
+    fileprivate func setupConstraints() {
```

Comment:
OMG Constraints... it's like Apple took revenge on devs who prefer writing custom views instead of using the Interface Builder. Now both methods are horrible. 😛

Reply:
We can always use frameworks like _SnapKit_ that make it more readable, but IMO is not worth.